compiler-rt/lib/scudo/standalone/combined.h

```diff
@@ -1504,6 +1529,28 @@ class Allocator {
       return;
     u32 AllocationRingBufferSize =
         static_cast<u32>(getFlags()->allocation_ring_buffer_size);
+    // We store alloc and free stacks for each entry.
+    constexpr auto kStacksPerRingBufferEntry = 2;
+    u32 TabSize = static_cast<u32>(roundUpPowerOfTwo(kStacksPerRingBufferEntry *
+                                                     AllocationRingBufferSize));
+    constexpr auto kFramesPerStack = 8;
+    static_assert(isPowerOfTwo(kFramesPerStack));
+    u32 RingSize = static_cast<u32>(TabSize * kFramesPerStack);
+    DCHECK(isPowerOfTwo(RingSize));
+    static_assert(sizeof(StackDepot) % alignof(atomic_u64) == 0);
+
+    StackDepotSize = sizeof(StackDepot) + sizeof(atomic_u64) * RingSize +
+                     sizeof(atomic_u32) * TabSize;
+    MemMapT DepotMap;
+    DepotMap.map(
+        /*Addr=*/0U, roundUp(StackDepotSize, getPageSizeCached()),
+        "scudo:stack_depot");
+    RawStackDepot = reinterpret_cast<char *>(DepotMap.getBase());
+    auto *Depot = reinterpret_cast<StackDepot *>(DepotMap.getBase());
+    Depot->init(RingSize, TabSize);
+    DCHECK(Depot->isValid(StackDepotSize));
```

Comment:
But init doesn't need to know the `StackDepotSize`.

Reply:
It seems to me that it's a logic conflict. Why do we init something with invalid arguments? Shouldn't we stop the initialization before we construct it?

Reply:
We don't, that's why the `DCHECK` is there? A `DCHECK` should assert what's always true.

Reply:
Sorry, I'm not saying how it's verified. I'm talking about the code structure. 

If the `DCHECK` fails, what will you review? It should be the `RingSize` and `TabSize` because they are used for the initialization. Therefore, checking the values before/in the middle of the initialization gives you the useful insight about what's going on. 

Initialization usually involves several steps, it'd be better to tell when there's something wrong instead of carrying the error to the end. If it causes any crash, that DCHECK won't help anything. Besides, at the end of initialization, I would expect that everything is fine. For example, I don't need to verify if `std::vector<int>(4, 0)` gives me 4 elements and all of them are zero.

IMO, the `DCHECK` is not used at the right place.

Reply:
I thought I replied to this, not sure if GitHub problem or I actually didn't:

I don't understand the problem with adding this DCHECK. This makes sure that the logic for `isValid` and for `init` are consistent, which is a good thing to check in debug builds. If it fails it's always something wrong with the code of either of them, because it should not fail by construction of the code.

> I don't need to verify if std::vector<int>(4, 0) gives me 4 elements and all of them are zero.

I don't think that's a correct analogy. If I was the implementer of std::vector then I might want to add some debug checks about the internal state of std::vector.

Reply:
Discussed offline, there was some misunderstanding, a user should _not_ be able to trigger this. But given we also assert this in tests, I just removed it.